src/daft-plan/src/optimization/rules/push_down_projection.rs

```diff
@@ -0,0 +1,398 @@
+use std::{collections::HashSet, convert::identity, sync::Arc};
+
+use common_error::DaftResult;
+
+use daft_core::schema::Schema;
+use daft_dsl::Expr;
+
+use crate::{
+    ops::{Aggregate, Project, Source},
+    LogicalPlan,
+};
+
+use super::{ApplyOrder, OptimizerRule};
+
+#[derive(Default)]
+pub struct PushDownProjection {}
+
+impl PushDownProjection {
+    pub fn new() -> Self {
+        Self {}
+    }
+
+    fn try_optimize_project(
+        &self,
+        projection: &Project,
+        plan: &LogicalPlan,
+    ) -> DaftResult<Option<Arc<LogicalPlan>>> {
+        let upstream_plan = projection.input.clone();
+        let upstream_schema = upstream_plan.schema();
+
+        // First, drop this projection if it is a no-op.
+        let projection_is_noop = {
+            let maybe_column_names = projection
+                .projection
+                .iter()
+                .map(|e| match e {
+                    Expr::Column(colname) => Some(colname.clone()),
+                    _ => None,
+                })
+                .collect::<Option<Vec<_>>>();
+
+            let all_columns_equal = maybe_column_names.map_or(false, |column_names| {
+                column_names
+                    .iter()
+                    .zip(upstream_schema.names().iter())
+                    .map(|(scol, pcol)| scol.as_ref() == pcol)
+                    .all(identity)
+            });
+
+            let no_unused_columns = upstream_schema.names().len() == projection.projection.len();
+
+            all_columns_equal && no_unused_columns
+        };
+        if projection_is_noop {
+            // Projection discarded but new root node has not been looked at;
+            // look at the new root node.
+            let new_plan = self.try_optimize(&upstream_plan)?.unwrap_or(upstream_plan);
+            return Ok(Some(new_plan));
+        }
```

Comment:
Another option that _might_ be cleaner is to have a separate `DropProjections` rule that looks for redundant/no-op projections, and we can order the `RuleBatch` such that `DropProjections` always runs after `PushDownProjection`, which I believe would eliminate the need to recursively call `self.try_optimize()` for some of the below match arms.